opal/corelib/runtime.js

```diff
@@ -996,6 +996,223 @@
   // Methods
   // -------
 
+  // Takes method arguments and extracts mlhs/restarg/kwarg/kwrest arguments
+  //
+  // @example
+  //   parameters = [1, 2, 3, 4, { d: 1, e: 2 }]
+  //   scheme = [
+  //     ['arg', 'a'],
+  //     ['restarg', 'b'],
+  //     ['arg', 'c'],
+  //     ['kwarg', 'd'],
+  //     ['kwrestarg', 'kwrest']
+  //   ]
+  //   (which is equivalent to (a, *b, c, d:, **kwrest))
+  //   Opal.scan_args(parameters, scheme)
+  //     // => { 'a': 1, 'b': [2, 3], 'c': 4, 'd': 1, 'kwrest': { 'e': 2 } }
+  //
+  // Scheme items must be the names of their sexp equivalents.
+  //
+  // :arg       is a      in "def m(a) end"
+  // :optarg    is b      in "def m(b=1) end"
+  // :restarg   is c      in "def m(*c) end"
+  // :mlhs      is (d, e) in "def m((d,e)) end"
+  // :kwarg     is :f     in "def m(f:) end"
+  // :kwoptarg  is :g     in "def m(g: 1) end"
+  // :kwrestarg is :f     in "def m(**f) end"
+  //
+  // @param parameters [Array] arguments that were passed to the method
+  // @param scheme [Array] array in format [ [<arg type>, <arg name>] ]
+  // @param expand [true, false] expands +parameters+ if it contains a single Array element and expand is +true+
+  Opal.scan_args = function(parameters, scheme, expand) {
+    if (expand && scheme.length > 1 && parameters.length == 1 && parameters[0].$$is_array) {
+      parameters = parameters[0];
+    }
+
+    var result = {},
+        extracted,
+        scheme_left = 0,
+        scheme_right = scheme.length - 1,
+        parameters_left = 0,
+        parameters_right = parameters.length - 1,
+        i;
+
+    // First, we should check for keyword arguments, as they are always in the end
+    var kwargs_done = false,
+        used_kwargs = {},
+        kwargs;
+
+    for (i = 0; i < scheme.length; i++) {
+      var argument_scheme = scheme[i],
+          argument_type = argument_scheme[0],
+          kwarg_name, kwarg_value;
+
+      if ((argument_type === 'kwarg' || argument_type === 'kwoptarg' || argument_type === 'kwrestarg') && !kwargs_done) {
+        kwargs = parameters[parameters_right];
+
+        if (kwargs != null && kwargs['$respond_to?']('to_hash', true)) {
+          kwargs = kwargs.$to_hash();
+          parameters_right--;
+        } else {
+          kwargs = {};
+        }
+        kwargs_done = true;
+      }
+
+      if (argument_type === 'kwarg') {
+        // extracting :kwarg
+        kwarg_name = argument_scheme[1];
+
+        if (!kwargs.$$is_hash || !kwargs.$$smap.hasOwnProperty(kwarg_name)) {
+          throw Opal.ArgumentError.$new("missing keyword: " + kwarg_name);
+        }
+        scheme_right--;
+        kwarg_value = kwargs.$$smap[kwarg_name];
+        used_kwargs[kwarg_name] = true;
+        result[kwarg_name] = kwarg_value;
+      } else if (argument_type === 'kwoptarg') {
```

Comment:
can you change this to:

```js
  …
}
else if (…
```

I'm more of a `} else` guy too, but it's for consistency with the rest of the code :)